### PR TITLE
perf(determinism): stop filling uninitialized memory under --deterministic-mode

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Clears `torch.utils.deterministic.fill_uninitialized_memory` under `--deterministic-mode`. Worth roughly **15% TFLOP/s** on large configs, with training bit-exact both run-to-run and across a resume.

Huge kudos to @nathan-nvidia for the insights and recommendation!

## What the fill is, and why reproducibility does not need it

Determinism costs an *independent output buffer*: a reduction that would otherwise accumulate into shared memory with unordered atomics writes into its own buffer instead, fixing the summation order. That is the mechanism that makes a run reproducible, and this PR does not touch it.

`torch.use_deterministic_algorithms(True)` also switches on a **separate** knob that fills every uninitialized allocation with NaN/MAX_INT, so a kernel *reading* memory it never wrote reads the same bytes each run.

```
  what makes a run reproducible          what the fill adds
  -----------------------------          ------------------
  independent output buffer              torch.empty        -> NaN/MAX_INT
  fixed reduction order                  empty_like         -> NaN/MAX_INT
  (kept)                                 empty_strided      -> NaN/MAX_INT
                                         Tensor.resize_     -> NaN/MAX_INT
                                         (only matters if something READS them)
```

With no padding feeding the computation, results are bit-identical either way — the fill only changes what an already-broken read would return.

## Why it is expensive

The cost is paid in kernel launches, not arithmetic: one extra fill kernel per empty allocation, serialized between real work.

```
  fill ON vs fill OFF, same configuration, nsys on one rank
  ---------------------------------------------------------
  kernel launches      up sharply; nearly all of the increase is FillFunctor
  kernel GPU time      barely moves
  wall time per step   nearly doubles
```

Kernel time flat while wall time doubles is the signature of GPU idle rather than slower kernels: the extra launches sit between real work and suppress the overlap between compute and communication. Gains are config-dependent — the more `torch.empty` calls a step makes, the larger.

## Correctness

Four legs in one allocation — train-through, a non-deterministic control, save at the midpoint, and resume from it — comparing tensorboard event files at full float32, which is the method the docs recommend:

| comparison | result |
|---|---|
| A/A (train-through vs an independent run over the same steps) | **all scalar points bit-identical** |
| RESUME (train-through vs resumed, over the post-resume steps) | **all scalar points bit-identical** |

Tags covered: `lm loss`, `grad-norm`, `num-zeros`, `params-norm`, `load_balancing_loss`, `seq_load_balancing_loss`, `learning-rate`, `loss-scale`, `batch-size`, and their `vs samples` variants.

## Documented: when padding is benign, and when it is not

Padding matters only if a computation reads it.

- **Benign — nothing reads it.** Computed values are bit-identical; only saved bytes differ.
- **Harmful — a computation consumes it.** A reduction over a padded tail, a GEMM with a rounded-up K, an unmasked attention region. Results then differ run to run, and the fill does not make them correct, only repeatably wrong.

Checkpoints are the benign case, and the one operational consequence of this PR: some saved tensors carry trailing pad slots no kernel writes, so two runs of the same configuration write files differing in those bytes while every trained value matches.

```
  sha256 of checkpoint shards   --> can report a mismatch it cannot explain
                                    (a hash cannot tell padding from divergence)
  tensorboard scalars @ float32 --> unaffected
                                    <- use this
```

The docs already recommend the scalar method; this PR adds a section on why the shard-hash method stops applying.

## Reverting locally

```python
import torch.utils.deterministic
torch.utils.deterministic.fill_uninitialized_memory = True
```

Worth doing when debugging a suspected uninitialized-memory read — that is the one thing the fill is good for, and turning it off is what surfaced the saved-padding case above.

## Issue tracking

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code
- [x] I have added relevant documentation
- [x] I have run the autoformatter.sh on my PR — reports `Changeset is empty, all good.`: the script scopes `CHANGED_FILES` to `megatron/core` and `tests/`, and this PR only touches `megatron/training/`.

### CODE REVIEW

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7093/commits/18278c3267f09726ba69b895d219797a43eb9ceb

**Review by jaredcasper — APPROVED** [review]

**Review by Connor-XY — APPROVED** [review]

**Comment by jaredcasper** [comment]
/ok to test [a8245e8](https://github.com/NVIDIA/Megatron-LM/pull/7093/commits/a8245e8ac6edeb27d464e19bdfae72b6fdf4259b)

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7093/commits/a40fdb21e5454a8c7fe4b91609de9753c8cd0711
